File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int maxValue;

    public int MaxPathSum(TreeNode root) {
        maxValue = int.MinValue;
        MaxPathSumTracker(root);
        return maxValue;
    }

    public int MaxPathSumTracker(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = Math.Max(MaxPathSumTracker(node.left), 0);
        int right = Math.Max(MaxPathSumTracker(node.right), 0);

        int sum = node.val + left + right;
        maxValue = Math.Max(maxValue, sum);

        return node.val + Math.Max(left, right);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_height = 0
        def recursiveDiameter(node):
            nonlocal max_height

            if node is None:
                return 0
            right = recursiveDiameter(node.right)
            left = recursiveDiameter(node.left)
            max_height = max(max_height, left + right)
        
            return 1 + max(left, right)
        recursiveDiameter(root)
        return max_height
        